### PR TITLE
[TACHYON-772] Enable Master Server PLAIN Authentication

### PR DESCRIPTION
**Pull request by shenguoquan**
JIRA: https://tachyon.atlassian.net/browse/TACHYON-772.

Enable Master Server PLAIN Authentication. I have written some integration tests for all kinds of authentication mode, such as NOSASL, SIMPLE, CUSTOM and KERBEROS.　There has some code about create a test loginUser for uniting test. I know @ooq has sent a PR about custom loginmodule. I will use his feature when it has been committed. @apc999 @ooq  please help me to review this PR. Thanks.

### CODE REVIEW

File: common/src/main/java/tachyon/security/LoginUser.java [hunk 1]
```diff
@@ -64,6 +64,19 @@ public static User get(TachyonConf conf) throws IOException {
...
+  public static void setLoginUser(User user) {
+    sLoginUser = user;
+  }
+
+  /**
+   * Create a user for testing Tachyon
+   * @param user the full user principal name
+   * @return a fake user for running unit tests
+   */
+  public static User createUserForTesting(String user) {
```
Comment (apc999) [inline common/src/main/java/tachyon/security/LoginUser.java]: Is it really necessary to have this method here?
Can we move it to unittests?

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+    mMasterInfo = mLocalTachyonCluster.getMasterInfo();
+    MasterClient masterClient =
+        new MasterClient(mMasterInfo.getMasterAddress(),
+            mExecutorService, mMasterTachyonConf);
+    Assert.assertFalse(masterClient.isConnected());
+    masterClient.connect();
+    Assert.assertTrue(masterClient.isConnected());
+    masterClient.user_createFile("/file-nosasl", "", Constants.DEFAULT_BLOCK_SIZE_BYTE, true);
+    Assert.assertTrue(masterClient.getFileStatus(-1, "/file-nosasl") != null);
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: use `assertNotNull` for simplicity? Similar comments applies below too.

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    Assert.assertTrue(masterClient.isConnected());
+    Assert.assertTrue(masterClient.getFileStatus(-1, "/file-custom") != null);
+    masterClient.close();
+
+    //stop cluster
+    mLocalTachyonCluster.stop();
+  }
+
+  @Test
+  public void customAuthenticationDenyConnectTest() throws Exception {
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: Can you also test the case where the username is not as expected and authentication fails?

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+    masterClient.close();
+
+    //stop cluster
+    mLocalTachyonCluster.stop();
+  }
+
+  @Test
+  public void simpleAuthenticationOpenCloseTest() throws Exception {
+    //simple authentication configure
+    setAuthenticationType(AuthType.SIMPLE);
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: It seems `simpleAuthenticationOpenCloseTest()` has exactly the same test and expected results as `noAuthenticationOpenCloseTest`. I thought simple authentication allows any combination of username/password, shall we test that (through a few combinations)?

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 5]
```diff
@@ -157,7 +159,7 @@ private void mkdir(String path) throws IOException {
     }
   }
 
-  public void start() throws IOException {
+  private void init() throws IOException {
```
Comment (apc999) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: is this change necessary?

File: servers/src/main/java/tachyon/master/TachyonMaster.java [hunk 6]
```diff
@@ -20,29 +20,32 @@
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 
+import javax.security.sasl.SaslException;
+
+import com.google.common.base.Preconditions;
```
Comment (apc999) [inline servers/src/main/java/tachyon/master/TachyonMaster.java]: This import order (com vs org) is inconsistent with [Tachyon code style](http://tachyon-project.org/Contributing-to-Tachyon.html#coding-style)

**Comment by apc999** [comment]
hi @shenguoquan , I uploaded some minor comments. Also please fix the checking failure.

Thanks

File: common/src/main/java/tachyon/security/LoginUser.java (hunk 1 above)
Comment (shenguoquan) [inline common/src/main/java/tachyon/security/LoginUser.java]: This methods was used to run unit tests with some specific login user. I know Qifan created a new feature TACHYON-786(Add CUSTOM_LOGIN loginmodule). I will use his feature when it has been committed. Thanks for your comments

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java (hunk 2 above)
Comment (shenguoquan) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: Thanks for your advice. I will change it.

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java (hunk 3 above)
Comment (shenguoquan) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: customAuthenticationDenyConnectTest has tested the connection failed case. It uses a NameMatchAuthenticationProvider authentication to verify only tahcyon user allowed to connect the Master service.

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java (hunk 4 above)
Comment (shenguoquan) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: Good advice. I have changed this code.

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java (hunk 5 above)
Comment (shenguoquan) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: It is necessary to change this code. Currently I can't set the configuration for tachyon cluster before it started. Take setting the authentication mechanism for example, when I want to set this configuration, I must at firstly get the tachyonConf insatnce. But the tachyonConf instance was created in the start method. That's the reason I changed this part source code

**Comment by shenguoquan** [comment]
@apc999 I have fixed the failure unit tests.

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by apc999** [comment]
Jenkins, test this please

**Comment by shenguoquan** [comment]
Hi @apc999 , It looks like my problem caused this failure. I will fix it. Thanks

**Comment by shenguoquan** [comment]
Hi @apc999 , I have fixed all of these failure UT.

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+    }
+  }
+
+  /**
+   * Test Tachyon client connects or disconnects to the Master. When the client connects
+   * successfully to the Master, it can successfully create file or not.
+   * @param filename
+   * @throws Exception
+   */
+  private void testAuthenticationOperation(String filename) throws Exception {
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: could you name this as `authenticationOperationTest` to stay consistent with the rest of the unittests?

Comment (shenguoquan) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: I will rename this method. THX

File: minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java [hunk 8]
```diff
@@ -72,7 +72,8 @@ public static void main(String[] args) throws Exception {
 
   private TachyonConf mWorkerConf;
 
-  public LocalTachyonCluster(long workerCapacityBytes, int quotaUnitBytes, int userBlockSize) {
+  public LocalTachyonCluster(long workerCapacityBytes, int quotaUnitBytes, int userBlockSize)
+      throws IOException {
```
Comment (ooq) [inline minicluster/src/main/java/tachyon/master/LocalTachyonCluster.java]: why adding this line?

File: servers/src/main/java/tachyon/master/TachyonMaster.java [hunk 9]
```diff
@@ -38,11 +38,12 @@
 import tachyon.Version;
 import tachyon.conf.TachyonConf;
 import tachyon.metrics.MetricsSystem;
+import tachyon.security.authentication.AuthenticationFactory;
 import tachyon.thrift.MasterService;
 import tachyon.underfs.UnderFileSystem;
 import tachyon.util.CommonUtils;
-import tachyon.util.network.NetworkAddressUtils;
 import tachyon.util.ThreadFactoryUtils;
+import tachyon.util.network.NetworkAddressUtils;
```
Comment (ooq) [inline servers/src/main/java/tachyon/master/TachyonMaster.java]: Can you check why this line shows up in the diff? They look identical and might embed some special character in either of them.

Comment (calvinjia) [inline servers/src/main/java/tachyon/master/TachyonMaster.java]: I think its because it was moved down one line.

Comment (ooq) [inline servers/src/main/java/tachyon/master/TachyonMaster.java]: ah :）then it's good, seems we put upper case names before lower case ones.

File: common/src/main/java/tachyon/security/LoginUser.java [hunk 10]
```diff
@@ -64,6 +64,19 @@ public static User get(TachyonConf conf) throws IOException {
     return sLoginUser;
   }
 
+  public static void setLoginUser(User user) {
```
Comment (ooq) [inline common/src/main/java/tachyon/security/LoginUser.java]: Javadoc? Is this also for testing only?

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    System.setProperty(Constants.TACHYON_SECURITY_AUTHENTICATION,
+        AuthType.CUSTOM.getAuthName());
+    //custom authenticationProvider configure
+    System.setProperty(Constants.TACHYON_AUTHENTICATION_PROVIDER_CUSTOM_CLASS,
+        NameMatchAuthenticationProvider.class.getName());
+    /**
+     * Using tachyon as loginUser for unit testing, only tachyon user is allowed to connect to
+     * Tachyon Master.
+     */
+    LoginUser.setLoginUser(LoginUser.createUserForTesting("tachyon"));
```
Comment (ooq) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: Is this line needed?

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+    //start cluster
+    mLocalTachyonCluster.start();
+
+    mMasterInfo = mLocalTachyonCluster.getMasterInfo();
+    // Using no-tachyon as loginUser to connect to Master, the IOException will be thrown
+    try {
+      mThrown.expect(IOException.class);
```
Comment (ooq) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: Any expected message?

**Comment by ooq** [comment]
@shenguoquan Thanks for the PR and apology for the delay. I left some comments.

**Comment by shenguoquan** [comment]
Depend on #1292

**Comment by ooq** [comment]
@shenguoquan Thanks for the info. #1292 is merged.

**Comment by shenguoquan** [comment]
This PR depends on #1412 .

**Comment by shenguoquan** [comment]
Hi @ooq , I have rebased the PR according to security branch. This PR depends on the #1412 . Please help me to review it. Thanks

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+     * Tachyon Master.
+     */
+    System.setProperty(Constants.TACHYON_SECURITY_USERNAME, "tachyon");
+    //start cluster
+    mLocalTachyonCluster.start();
+
+    mMasterInfo = mLocalTachyonCluster.getMasterInfo();
+    // Using no-tachyon as loginUser to connect to Master, the IOException will be thrown
+    clearLoginUser();
+    mThrown.expect(IOException.class);
```
Comment (ooq) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: Any expected message here?

**Comment by ooq** [comment]
Looks good to me except one minor change is needed.

File: integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java (hunk 13 above)
Comment (shenguoquan) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: Because the real exception was wrapped by the IOException from MasterClient. So I can't test the exception message here. Thanks for your comment.

Comment (shenguoquan) [inline integration-tests/src/test/java/tachyon/master/MasterClientAuthenticationIntegrationTest.java]: The MasterClient will  throw new IOException("Failed to connect with master @ " + mMasterAddress + " after " + (retry.getRetryCount()) + " attempts", lastException); So I think the expect message couldn't be tested.

**Comment by apc999** [comment]
+1 LGTM
